rust/src/mime/mime.rs

```diff
@@ -0,0 +1,581 @@
+                if let Ok((rem, eof)) = mime_consume_chunk(&ctx.boundary, input) {
+                    consumed += (input.len() - rem.len()) as u32;
+                    if eof {
+                        ctx.state = MimeParserState::Start;
+                        ctx.filename.clear();
+                        return (MimeParserResult::MimeFileClose, consumed, warnings);
+                    } else {
+                        // + 2 for \r\n
+                        if rem.len() < ctx.boundary.len() + 2 {
+                            return (MimeParserResult::MimeFileChunk, consumed, warnings);
+                        }
+                        input = rem;
+                    }
+                } else {
+                    return (MimeParserResult::MimeNeedsMore, consumed, warnings);
+                }
+            }
+        }
+    }
+    return (MimeParserResult::MimeNeedsMore, consumed, warnings);
+}
+
+pub fn mime_state_init(i: &[u8]) -> Option<MimeStateHTTP> {
+    let mut sections_values = Vec::new();
+    if let Some(value) = mime_find_header_token(i, "boundary".as_bytes(), &mut sections_values) {
+        if value.len() <= RS_MIME_MAX_TOKEN_LEN {
+            let mut r = MimeStateHTTP {
+                boundary: Vec::with_capacity(2 + value.len()),
+                ..Default::default()
+            };
+            // start wih 2 additional hyphens
+            r.boundary.push(b'-');
+            r.boundary.push(b'-');
+            for c in value {
+                // unescape
+                if *c != b'\\' {
+                    r.boundary.push(*c);
+                }
+            }
+            return Some(r);
+        }
+    }
+    return None;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCMimeStateInit(input: *const u8, input_len: u32) -> *mut MimeStateHTTP {
+    let slice = build_slice!(input, input_len as usize);
+
+    if let Some(ctx) = mime_state_init(slice) {
+        let boxed = Box::new(ctx);
+        return Box::into_raw(boxed) as *mut _;
+    }
+    return std::ptr::null_mut();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCMimeParse(
+    ctx: &mut MimeStateHTTP, input: *const u8, input_len: u32, consumed: *mut u32,
+    warnings: *mut u32,
+) -> MimeParserResult {
+    let slice = build_slice!(input, input_len as usize);
+    let (r, c, w) = mime_process(ctx, slice);
+    *consumed = c;
+    *warnings = w;
+    return r;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCMimeStateGetFilename(
+    ctx: &mut MimeStateHTTP, buffer: *mut *const u8, filename_len: *mut u16,
+) {
+    if !ctx.filename.is_empty() {
+        *buffer = ctx.filename.as_ptr();
+        if ctx.filename.len() < u16::MAX.into() {
+            *filename_len = ctx.filename.len() as u16;
+        } else {
+            *filename_len = u16::MAX;
+        }
+    } else {
+        *buffer = std::ptr::null_mut();
+        *filename_len = 0;
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCMimeStateFree(ctx: &mut MimeStateHTTP) {
+    // Just unbox...
```

Comment:
I don't think this is new code, but these comments are wrong now. We're explicitly dropping, not just unboxing.

Reply:
ok